### DISCUSSION TITLE
Possible feature requests for Py5Vector...

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#242] [Ideas]
In Processing you can set the magnitude of a vector to a negative scalar... this will "flip" the heading...

![image](https://user-images.githubusercontent.com/3694604/222146965-3021a9e4-cd40-4b64-8aa6-f4f62c1f90ea.png)

Yesterday at Noite de Processing we were jamming and live coding a "ParticleLife" that was handy...

My solution was:

```python
    v = Py5Vector(100, 100)
    new_mag = -10
    if new_mag < 0:
        v =* -1
        new_mag = -new_mag
    v.mag = new_mag
    print(v)
```

Could there be a less costly solution inside Py5Vector?

Also Processing "fails silently" to set magnitude to a "null vector". Should we follow that?

![image](https://user-images.githubusercontent.com/3694604/222148868-39e5e038-fd53-4556-b606-b0b490d6648f.png)

**Comment by hx2A** [comment]
> In Processing you can set the magnitude of a vector to a negative scalar... this will "flip" the heading...

That's interesting! Right now the Py5Vector will reject a negative value for the `set_mag()` method. I think we should allow negative values, like Processing does. This is an easy change.

> Also Processing "fails silently" to set magnitude to a "null vector". Should we follow that?

I did some experimenting and I see that this is valid Processing code:

```java
PVector v = new PVector();
v.normalize();
println(v);
// [ 0.0, 0.0, 0.0 ]
```

py5 would reject the similar code, not allowing you to normalize a vector with a magnitude of zero. This does bother me a bit, but being consistent with Processing is probably the better choice. And if this is allowed in py5, using `set_mag()` on a zero vector will also be allowed.

**Reply by villares** [reply]
How about emitting a warning in the zero magnitude vector case? It could help someone with debugging perhaps?

**Reply by hx2A** [reply]
> How about emitting a warning in the zero magnitude vector case? It could help someone with debugging perhaps?

Yes, that's a good idea. And I'll use Python's builtin warnings library to do that. py5 is already using it in a few places. It contains features that let you suppress warning messages if you don't want to be told the same thing over and over.
